server/src/pages/shot/text-tool.js

```diff
@@ -0,0 +1,278 @@
+
+let isMousedown;
+let mousedownPosition;
+let mousedownSelection;
+
+exports.TextTool = class TextTool extends React.Component {
+  constructor(props) {
+    super(props);
+    this.canvas = React.createRef();
+    this.el = React.createRef();
+
+    this.canvasCssWidth = props.canvasCssWidth;
+    this.canvasCssHeight = props.canvasCssHeight;
+
+    const INIT_LEFT = this.canvasCssWidth / 2;
+    const INIT_TOP = this.canvasCssHeight / 2;
+
+    const FONT_SIZE = this.getFontSize(this.props.textSize) || INIT_FONT_SIZE;
+
+    const textSelection = new Selection(INIT_LEFT,
+                                        INIT_TOP,
+                                        INIT_LEFT + INIT_WIDTH,
+                                        FONT_SIZE * 1.3 + INIT_TOP + TEXT_INPUT_PADDING);
+
+    this.state = {
+      textSelection,
+      textContent: "",
+      color: this.props.color,
+      colorName: this.props.colorName,
+      textSize: this.props.textSize
+    };
+
+  }
+
+  getFontSize(textSize) {
+    if (!textSize) {
+      return null;
+    }
+    const arrSizes = {"smalll": 16, "medium": 24, "large": 36};
+    return arrSizes[textSize];
+  }
+
+  componentDidMount() {
+    // Initalize drawing context when user first clicks on canvas
+    this.drawingContext = this.canvas.current.getContext("2d");
+    this.drawingContext.scale(this.props.canvasPixelRatio, this.props.canvasPixelRatio);
+    this.drawingContext.textBaseline = "top";
+
+    this.el.current.focus();
+
+    if (this.props.toolbarOverrideCallback) {
+      this.props.toolbarOverrideCallback();
+    }
+  }
+
+  componentWillUnmount() {
+    if (this.props.toolbarOverrideCallback) {
+      this.props.toolbarOverrideCallback();
+    }
+  }
+
+  render() {
+    const textSelectionBox = this.renderTextBox();
+    return <div
+      className={`text-tool-container centered`}
+      onMouseDown={this.onMouseDown.bind(this)}
+      style={{height: this.props.baseCanvas.style.height, width: this.props.baseCanvas.style.width}}>
+      {textSelectionBox}
+    </div>;
+  }
+
+  renderTextBox() {
+    if (!this.state.textSelection) {
+      return null;
+    }
+
+    const selectionTopPx = `${this.state.textSelection.top}px`;
+    const selectionLeftPx = `${this.state.textSelection.left}px`;
+
+    const selectedStyles = {
+      position: "absolute",
+      top: selectionTopPx,
+      left: selectionLeftPx
+    };
+
+    return [
+      <canvas ref={this.canvas} id="text-canvas" key="base" width={this.props.baseCanvas.width}
+        height={this.props.baseCanvas.height}
+        style={{width: this.canvasCssWidth,
+                height: this.canvasCssHeight}}>
+      </canvas>,
+
+      <div id="text-input" ref={this.el} contentEditable="true" key="text" onDoubleClick={this.editText.bind(this)} onKeyUp={this.drawText.bind(this)}
+         className={`${this.state.textSize} ${this.state.colorName} text`} style={selectedStyles}>
+      </div>
+    ];
+  }
+
+  renderToolbar() {
```

Comment:
The toolbar is rendered by the editor, and the text tool hasn't been updated at that time.  That's why the selected text size is always one behind.  You could try using `componentDidUpdate` to trigger an editor re-render when the text tool's text size has changed.

Reply:
chenba: I have updated PR to keep textSize state inside textTool. Need more clarity on how componentDidUpdate works and how to trigger an editor componentDidUpdate from onChange event in TextTool

Reply:
A react component re-renders when its state changes.  We need the editor to re-render after the text tool has updated, and we can do that by changing the editor's state with a callback.  In the text tool's `componentDidUpdate` (https://reactjs.org/docs/react-component.html#componentdidupdate) you can determine if the `textSize` has changed and call the callback.

This is basically how `overrideToolbar` works; feel free to update it to fit your need.

Reply:
perfect makes sense, thanks!